docs/asdf/developer_versioning.rst

```diff
@@ -0,0 +1,253 @@
+that it supports.  It also provides any custom serialization/deserialization
+behavior that is required -- AsdfType provides a default
+implementation that is only able to get and set attributes on dict-like
+objects.
+
+In some cases, the AsdfType subclass also serves as the deserialized
+object type.  For example, ``asdf.types.core.Software`` subclasses both
+AsdfType and dict.  Its AsdfType-like behavior is
+to identify its schema and version, while its dict-like behavior is
+to act as a container for the attributes described by the schema.  The class
+definition is mostly empty because as a dict it can rely on
+AsdfType's default implementation for (de)serialization.
+
+Meanwhile, other AsdfType subclasses deserialize ASDF objects
+into instances of entirely separate classes.  For example,
+``asdf.types.core.complex.ComplexType`` handles complex number types,
+which aren't natively supported by YAML.  ComplexType includes
+an additional class attribute, ``types``, that lists the types that
+it is able to handle.  It also provides custom implementations
+of the ``to_tree`` and ``from_tree`` class methods, which enable it to
+serialize a complex value into the appropriate string, and later
+rebuild the complex value from that string.  This additional code is
+necessary because ComplexType does not (de)serialize itself.
+
+We won't find an explicit list of AsdfType subclasses
+in the code; that list is assembled at runtime by AsdfType's
+metaclass, ``asdf.types.AsdfTypeMeta``.  The list can be inspected in
+the console like so:
+
+.. code-block:: pycon
+
+    >>> import asdf
+    >>> asdf.types._all_asdftypes # doctest: +SKIP
+    ...
+
+The AsdfType class attributes relevant to versioning are as follows:
+
+- *name*: the base name of the schema, without its version string.
+  For example, a schema located at
+  ``asdf-standard/schemas/stsci.edu/asdf/core/example-1.2.0.yaml`` will
+  have a name value of ``"core/example"``.
+
+- *version*: the primary schema version supported by the AsdfType.
+  For the example above, version should be set to ``"1.2.0"``.  This should
+  be the latest version that the AsdfType supports.
+
+- *supported_versions*: a set of schema versions that the AsdfType
+  supports.  In the above example, this might be
+  ``{"1.0.0", "1.1.0", "1.2.0"}``.
+
+AsdfType selection rules
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+On read, the library will ideally be able to identify an AsdfType
+subclass that explicitly supports a given tag (either in the ``version``
+class attribute or ``supported_versions``.  If that is not possible,
+it proceeds as follows:
+
+- Use the AsdfType that supports the latest version that is
+  less than the tag version.  For example, if the tag is example-1.2.0,
+  and AsdfType are available for 1.1.0 and 1.3.0, it will
+  use the 1.1.0 subclass.
+- If the above fails, use the earliest available AsdfType
+
+The rules for selecting an AsdfType for a given tag are implemented
```

Comment:
Might be good to summarize when and if warnings are issued when exact matches are not found and it assumes the earliest.

Reply:
That's a good idea, I don't know the answer but I'll look into it

Reply:
(added documentation on the warnings)